xaitk_saliency/utils/gen_coco_sal.py

```diff
@@ -0,0 +1,262 @@
+        """
+
+        if verbose:
+            _log: Callable[[str], None] = LOG.info
+        else:
+            _log = _mute_log
+
+        num_imgs = len(dets_dset.imgs)
+
+        sal_maps = {}
+        for img_i, (img_id, det_ids) in enumerate(dets_dset.gid_to_aids.items()):
+
+            num_dets = len(det_ids)
+            # continue if there are no dets for this image
+            if num_dets == 0:
+                continue
+
+            img_file = dets_dset.get_image_fpath(img_id)
+            ref_img = np.asarray(Image.open(img_file))
+
+            ref_dets_mat = _coco_dets_to_formatted_mat(dets_dset, det_ids)
+
+            _log(f'[{img_id}] ({img_i+1}/{num_imgs}) {num_dets} reference detections found.')
+
+            _log(f'[{img_id}] ({img_i+1}/{num_imgs}) Perturbing image.')
+
+            pert_masks = img_perturber(ref_img)
+            pert_imgs = occlude_image_batch(ref_img, pert_masks)
+
+            _log(f'[{img_id}] ({img_i+1}/{num_imgs}) Getting perturbation detections.')
+
+            pert_dets_iter = detector(pert_imgs)
+
+            _log(f'[{img_id}] ({img_i+1}/{num_imgs}) Generating saliency maps.')
+
+            pert_dets_mat = _dets_to_formatted_mat(pert_dets_iter, dets_dset.cats)
+
+            img_sal_maps = sal_generator(ref_dets_mat, pert_dets_mat, pert_masks)
+
+            for i, det_id in enumerate(det_ids):
+                sal_maps[det_id] = img_sal_maps[i]
+
+        return sal_maps
+
+    def _name_to_cat_id(
+        name: str,
+        cats: dict
+    ) -> int:
+        """
+        Converts category name to id.
+
+        :param name: Category name.
+        :param cats: Dictionary of categories, as taken from the "categories"
+            section of a COCO style annotation dictionary.
+        :returns: Id of matching category, or -1 if no matching category is
+            found.
+        """
+        for cat_id in cats:
+            if name == cats[cat_id]['name']:
+                return cat_id
+
+        return -1
+
+    def _map_score_dict_keys(
+        score_keys: KeysView,
+        coco_cats: dict
+    ) -> Dict[Hashable, int]:
+        """
+        Matches score dict keys, as returned by an implementation of
+        ``DetectImageObjects``, with categories from a COCO dataset.
+
+        :param score_keys: List of key values from score dict, as returned by
+            an implementation of ``DetectImageObjects``.
+        :param coco_cats: Dict of categories, as taken from the "categories"
+            section of a COCO style annotation dictionary.
+        :returns: Dictionary mapping score dict key to category id.
+        """
+
+        cat_map = {}  # type: Dict[Hashable, int]
+        for key in score_keys:
+            if type(key) is int or type(key) is float or (type(key) is str and key.isnumeric()):
+                # assume key is a cat id
+                if int(key) in coco_cats:
+                    cat_map[key] = coco_cats[key]['id']
```

Comment:
2 things:
1) I think this will fail if `type(key) is not int`. You are casting `key` to int before checking containment and then accessing with the original type.
2) Wouldn't `coco_cats[key]['id']` be the same as `int(key)`? If so, the assignment could just be of `int(key)` to save on total function calls:
```suggestion
                key_int = int(key)
                if key_int in coco_cats:
                    cat_map[key] = key_int
```